### ISSUE TITLE
Unusual memory consumption with minio distributed server (XL) 

### ISSUE THREAD

**Issue by vadmeste** [minio/minio#3403]
@dcharbonnier:
GET and PUT during at least one hour, all the bandwidth that the servers can consume, with ~150 clients connected to 2 clusters of 4 machines, between 500Mbits and 1Gbit and it's a lot of read/write with files of ~50Mo

## Expected Behavior
Normal execution

## Current Behavior
OOM killer killing Minio

## Your Environment
All

**Comment by vadmeste** [comment]
@dcharbonnier, are you already balancing connections? approximately how many clients are concurrently connected to one node in your cluster ?

**Comment by harshavardhana** [comment]
```
[1459571.159608] [ pid ]   uid  tgid total_vm      rss nr_ptes nr_pmds swapents oom_score_adj name
[1459571.159614] [  296]     0   296    12934        1      31       3       80             0 systemd-journal
[1459571.159617] [  334]     0   334    11180       12      22       3      303         -1000 systemd-udevd
[1459571.159620] [  602]   106   602    10725        3      26       3      119          -900 dbus-daemon
[1459571.159622] [  608]   104   608    64099       81      27       3      224             0 rsyslogd
[1459571.159624] [  609]     0   609    69056       45      35       4      262             0 accounts-daemon
[1459571.159627] [  611]     0   611     7137        6      18       3       81             0 systemd-logind
[1459571.159629] [  617]     0   617     7241        7      20       3       65             0 cron
[1459571.159631] [  658]     0   658     3381       27      11       3       51             0 mdadm
[1459571.159633] [  692]     0   692     4868       28      14       3       33             0 irqbalance
[1459571.159635] [  771]     0   771     4030        0      11       3      216             0 dhclient
[1459571.159638] [  810]   108   810   107587        0      59       3     3161             0 named
[1459571.159640] [  817]     0   817    16380       12      36       3      169         -1000 sshd
[1459571.159642] [  829]     0   829     3224        0      11       3       54             0 ntpd
[1459571.159644] [  830]   109   830     3262        0      12       3       81             0 ntpd
[1459571.159646] [  831]   109   831     4303        1      14       3       63             0 ntpd
[1459571.159648] [  836]     0   836     3974        0      13       3       40             0 agetty
[1459571.159651] [30306]     0 30306   225425       31      87       6     6297          -500 dockerd
[1459571.159653] [30340]     0 30340   112458       15      34       6     1415          -500 docker-containe
[1459571.159656] [30502]     0 30502    85392        0      24       5      895          -500 docker-containe
[1459571.159658] [30517]     0 30517  8324301  8109022   16261      37   195938             0 minio
[1459571.159660] [30974]     0 30974    87025        0      25       6     1383          -500 docker-containe
[1459571.159662] [30989]     0 30989   151016     5837      55      27      349             0 node
[1459571.159665] [22683]     0 22683    23842      113      51       3      122             0 sshd
[1459571.159667] [22685]  1000 22685    11342       51      25       3      140             0 systemd
[1459571.159669] [22686]  1000 22686    35765      109      35       3      373             0 (sd-pam)
[1459571.159671] [22706]  1000 22706    23842      127      48       3      112             0 sshd
[1459571.159673] [22709]  1000 22709     5730      205      16       3       75             0 nload
```

**Comment by harshavardhana** [comment]
```
docker exec c4ca8d9bfdf3 minio version
Version: 2016-11-30T10:03:25Z
Release-Tag: DEVELOPMENT.2016-11-30T10-03-25Z
Commit-ID: bc9509bc8a95cc35113afe850428bee0d16ff056

[1459571.159445] minio invoked oom-killer: gfp_mask=0x24201ca, order=0, oom_score_adj=0
[1459571.159449] minio cpuset=c4ca8d9bfdf39a2dd0ee0b1d1d5d0b285a76a8f02f273e2adb40ff32df37df25 mems_allowed=0
[1459571.159455] CPU: 1 PID: 30533 Comm: minio Not tainted 4.4.0-47-generic #68-Ubuntu
[1459571.159456] Hardware name: Dell Inc. PowerEdge R210 II/03X6X0, BIOS 2.6.1 09/13/2013
[1459571.159459]  0000000000000286 00000000a6fd23de ffff880815c5fa30 ffffffff813f5aa3
[1459571.159462]  ffff880815c5fc28 ffff88081a81b700 ffff880815c5faa0 ffffffff81209ebe
[1459571.159464]  0000000000000015 0000000000000000 ffff8800b9827780 ffff8808157f6040
[1459571.159467] Call Trace:
[1459571.159475]  [<ffffffff813f5aa3>] dump_stack+0x63/0x90
[1459571.159479]  [<ffffffff81209ebe>] dump_header+0x5a/0x1c5
[1459571.159482]  [<ffffffff8138ed24>] ? apparmor_capable+0xc4/0x1b0
[1459571.159486]  [<ffffffff81191ab2>] oom_kill_process+0x202/0x3c0
[1459571.159488]  [<ffffffff81191ed9>] out_of_memory+0x219/0x460
[1459571.159492]  [<ffffffff81197f20>] __alloc_pages_nodemask+0x9d0/0xb60
[1459571.159497]  [<ffffffff811e18ac>] alloc_pages_current+0x8c/0x110
[1459571.159500]  [<ffffffff8118e35b>] __page_cache_alloc+0xab/0xc0
[1459571.159503]  [<ffffffff8119058a>] filemap_fault+0x14a/0x3f0
[1459571.159506]  [<ffffffff812a11b6>] ext4_filemap_fault+0x36/0x50
[1459571.159510]  [<ffffffff811bcfd0>] __do_fault+0x50/0xe0
[1459571.159514]  [<ffffffff811c0adb>] handle_mm_fault+0xf8b/0x1820
[1459571.159518]  [<ffffffff8106b4e7>] __do_page_fault+0x197/0x400
[1459571.159520]  [<ffffffff8106b772>] do_page_fault+0x22/0x30
[1459571.159525]  [<ffffffff81837178>] page_fault+0x28/0x30
[1459571.159527] Mem-Info:
[1459571.159531] active_anon:7536369 inactive_anon:584159 isolated_anon:0
                  active_file:1602 inactive_file:2395 isolated_file:0
                  unevictable:0 dirty:0 writeback:0 unstable:0
                  slab_reclaimable:9098 slab_unreclaimable:7130
                  mapped:342 shmem:3306 pagetables:17045 bounce:0
                  free:50966 free_pcp:346 free_cma:0
[1459571.159536] Node 0 DMA free:15900kB min:32kB low:40kB high:48kB active_anon:0kB inactive_anon:0kB active_file:0kB inactive_file:0kB unevictable:0kB isolated(anon):0kB isolated(file):0kB present:15984kB managed:15900kB mlocked:0kB dirty:0kB writeback:0kB mapped:0kB shmem:0kB slab_reclaimable:0kB slab_unreclaimable:0kB kernel_stack:0kB pagetables:0kB unstable:0kB bounce:0kB free_pcp:0kB local_pcp:0kB free_cma:0kB writeback_tmp:0kB pages_scanned:0 all_unreclaimable? yes
[1459571.159541] lowmem_reserve[]: 0 2911 32093 32093 32093
[1459571.159544] Node 0 DMA32 free:122656kB min:6128kB low:7660kB high:9192kB active_anon:2295932kB inactive_anon:584236kB active_file:0kB inactive_file:144kB unevictable:0kB isolated(anon):0kB isolated(file):0kB present:3098848kB managed:3018276kB mlocked:0kB dirty:0kB writeback:0kB mapped:4kB shmem:13148kB slab_reclaimable:3632kB slab_unreclaimable:2352kB kernel_stack:144kB pagetables:3148kB unstable:0kB bounce:0kB free_pcp:436kB local_pcp:120kB free_cma:0kB writeback_tmp:0kB pages_scanned:1116 all_unreclaimable? yes
[1459571.159550] lowmem_reserve[]: 0 0 29182 29182 29182
[1459571.159553] Node 0 Normal free:65308kB min:61420kB low:76772kB high:92128kB active_anon:27849544kB inactive_anon:1752400kB active_file:6408kB inactive_file:9436kB unevictable:0kB isolated(anon):0kB isolated(file):0kB present:30408704kB managed:29882468kB mlocked:0kB dirty:0kB writeback:0kB mapped:1364kB shmem:76kB slab_reclaimable:32760kB slab_unreclaimable:26168kB kernel_stack:3408kB pagetables:65032kB unstable:0kB bounce:0kB free_pcp:948kB local_pcp:120kB free_cma:0kB writeback_tmp:0kB pages_scanned:96164 all_unreclaimable? yes
[1459571.159558] lowmem_reserve[]: 0 0 0 0 0
[1459571.159561] Node 0 DMA: 1*4kB (U) 1*8kB (U) 1*16kB (U) 0*32kB 2*64kB (U) 1*128kB (U) 1*256kB (U) 0*512kB 1*1024kB (U) 1*2048kB (M) 3*4096kB (M) = 15900kB
[1459571.159572] Node 0 DMA32: 109*4kB (UME) 219*8kB (UME) 249*16kB (UME) 238*32kB (UME) 140*64kB (UME) 86*128kB (UME) 66*256kB (UME) 56*512kB (UME) 31*1024kB (UME) 4*2048kB (E) 1*4096kB (E) = 123356kB
[1459571.159584] Node 0 Normal: 406*4kB (UME) 300*8kB (UME) 432*16kB (UE) 393*32kB (UME) 198*64kB (UME) 108*128kB (UME) 29*256kB (UME) 9*512kB (ME) 3*1024kB (M) 0*2048kB 0*4096kB = 65112kB
[1459571.159595] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=2048kB
[1459571.159597] 8307 total pagecache pages
[1459571.159598] 874 pages in swap cache
[1459571.159600] Swap cache stats: add 406404, delete 405530, find 12817/14477
[1459571.159601] Free swap  = 0kB
[1459571.159602] Total swap = 1048568kB
[1459571.159603] 8380884 pages RAM
[1459571.159604] 0 pages HighMem/MovableOnly
[1459571.159605] 151723 pages reserved
[1459571.159606] 0 pages cma reserved
[1459571.159607] 0 pages hwpoisoned
[1459571.159608] [ pid ]   uid  tgid total_vm      rss nr_ptes nr_pmds swapents oom_score_adj name
[1459571.159614] [  296]     0   296    12934        1      31       3       80             0 systemd-journal
[1459571.159617] [  334]     0   334    11180       12      22       3      303         -1000 systemd-udevd
[1459571.159620] [  602]   106   602    10725        3      26       3      119          -900 dbus-daemon
[1459571.159622] [  608]   104   608    64099       81      27       3      224             0 rsyslogd
[1459571.159624] [  609]     0   609    69056       45      35       4      262             0 accounts-daemon
[1459571.159627] [  611]     0   611     7137        6      18       3       81             0 systemd-logind
[1459571.159629] [  617]     0   617     7241        7      20       3       65             0 cron
[1459571.159631] [  658]     0   658     3381       27      11       3       51             0 mdadm
[1459571.159633] [  692]     0   692     4868       28      14       3       33             0 irqbalance
[1459571.159635] [  771]     0   771     4030        0      11       3      216             0 dhclient
[1459571.159638] [  810]   108   810   107587        0      59       3     3161             0 named
[1459571.159640] [  817]     0   817    16380       12      36       3      169         -1000 sshd
[1459571.159642] [  829]     0   829     3224        0      11       3       54             0 ntpd
[1459571.159644] [  830]   109   830     3262        0      12       3       81             0 ntpd
[1459571.159646] [  831]   109   831     4303        1      14       3       63             0 ntpd
[1459571.159648] [  836]     0   836     3974        0      13       3       40             0 agetty
[1459571.159651] [30306]     0 30306   225425       31      87       6     6297          -500 dockerd
[1459571.159653] [30340]     0 30340   112458       15      34       6     1415          -500 docker-containe
[1459571.159656] [30502]     0 30502    85392        0      24       5      895          -500 docker-containe
[1459571.159658] [30517]     0 30517  8324301  8109022   16261      37   195938             0 minio
[1459571.159660] [30974]     0 30974    87025        0      25       6     1383          -500 docker-containe
[1459571.159662] [30989]     0 30989   151016     5837      55      27      349             0 node
[1459571.159665] [22683]     0 22683    23842      113      51       3      122             0 sshd
[1459571.159667] [22685]  1000 22685    11342       51      25       3      140             0 systemd
[1459571.159669] [22686]  1000 22686    35765      109      35       3      373             0 (sd-pam)
[1459571.159671] [22706]  1000 22706    23842      127      48       3      112             0 sshd
[1459571.159673] [22709]  1000 22709     5730      205      16       3       75             0 nload
[1459571.159675] Out of memory: Kill process 30517 (minio) score 979 or sacrifice child
[1459571.159682] Killed process 30517 (minio) total-vm:33297204kB, anon-rss:32436088kB, file-rss:0kB
```

**Comment by dcharbonnier** [comment]
@harshavardhana same crash with 
```
Version: 2016-12-07T09:00:12Z
Release-Tag: DEVELOPMENT.2016-12-07T09-00-12Z
Commit-ID: 64b942e34c973cdc77474e7137c8429c6b987a0c
```
On a 32Go ram machine

was you able to reproduce it with this version ?

**Comment by dcharbonnier** [comment]
@vadmeste sorry, I didn't reply, ~200 clients, 2 minio clusters of 4 machines each, sharding on the 2 clusters, round-robin on the hosts of the cluster

**Comment by dcharbonnier** [comment]
New test, new crash with a stack
https://gist.github.com/dcharbonnier/146081d24eba4b9787da5ea3765deb04

**Comment by harshavardhana** [comment]
Fixed by #3495
